docs/rfcs/0002-browser-search-rewrite.md

```diff
@@ -0,0 +1,127 @@
+
+Since the state will move to `browser-state`, additional functionality (querying suggestions, storage, middleware) can move to `feature-search`. This will make `browser-search` obsolete. With that introducing a `concept-search` is not required. An app can use a custom implementation by replacing the middleware, storage or handling search state completely differently.
+
+## Reference-level explanation
+
+The new implementation will use `browser-state` to model all search state.
+
+Functionality on top of the state (querying suggestions, storage, middleware) will live in `feature-search`. With that `browser-search` will no longer be needed and can stay around as-is until no longer used by any app.
+
+### State
+
+`BrowserState` will get a new `search` property with `SearchState` type that lives in `browser-state`:
+
+```Kotlin
+data class BrowserState(
+   // ...
+   val search: SearchState
+)
+
+/**
+ * Value type that represents the state of available search engines.
+ *
+ * @property searchEngines List of loaded search engines.
+ * @property defaultSearchEngineId ID of the default search engine.
+ */
+data class SearchState(
+  val searchEngines: List<SearchEngine>,
+  val defaultSearchEngineId: String
+)
+```
+
+`SearchState` will contain all `SearchEngine`s (bundled and custom). Additional extension methods will make it easier to select specific subsets or the default `SearchEngine`.
+
+`SearchEngine` will be turned into a data class and moved to `browser-store`. Other methods like `buildSearchUrl()` will be implemented as extension methods.
+
+### Storage
+
+There will be two storage classes:
+
+* `SearchEngineStorage`: A persistent storage for `SearchEngine`s. The primary use case is for persisting custom search engines added by users.
+* `SearchEngineDefaults`: A provider of a list of default search engines (for the user's region) based on `list.json` and the search plugins currently included in `browser-search`.
+
+Those storage classes will have visibility `internal` and will not be used by the app directly.
+
+All storage access methods will be suspending functions to avoid thread-blocking access:
+
+```Kotlin
+internal class SearchEngineStorage {
+    suspend fun getSearchEngines() = withContext(Dispatchers.IO) {
+      // ...
+    }
+}
+```
+
+### Middleware
+
+A `SearchMiddleware` that will be installed on `BrowserStore` will be responsible for:
+
+* Loading the initial set of search engines (from the two storages above) and adding them to `BrowserState` by dispatching actions on `BrowserStore`.
+* Persisting state changes it observes in the specific storage:
+  * Adding search engines
+  * Removing search engines
+  * Changing the default search engine
+* Updating search engines at runtime (e.g. region or locale change)
```

Comment:
Region change from MLS?
How would this work?
Polling at regular intervals seems rather wasteful.
Maybe this could be tried once per app start?

Could updating search engines change even the default currently used by the user?

Reply:
This is pretty much undefined at this point. The primary point here is that the middleware will take care of something like that and we will not pass that burden onto the app.

* Issue #1720 is about looking at more data sources than only MLS (or locale). In that same bug Mike Connor (who from our point of view owns the search engine plugins and region mapping) confirmed that we are in general okay with switching defaults. Of course that would need further clarification of how and when that would happen.
* As far as I know the desktop version will soon roll out a change that will make it re-request and in certain situation change its region.